src/Illuminate/Foundation/Defer/DeferredCallbackCollection.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+namespace Illuminate\Foundation\Defer;
+
+use ArrayAccess;
+use Closure;
+use Countable;
+use Illuminate\Support\Collection;
+
+class DeferredCallbackCollection implements ArrayAccess, Countable
+{
+    /**
+     * All of the deferred callbacks.
+     *
+     * @var array
+     */
+    protected array $callbacks = [];
+
+    /**
+     * Get the first callback in the collection.
+     *
+     * @return callable
+     */
+    public function first()
+    {
+        return array_values($this->callbacks)[0];
+    }
+
+    /**
+     * Invoke the deferred callbacks.
+     *
+     * @return void
+     */
+    public function invoke(): void
+    {
+        $this->invokeWhen(fn () => true);
+    }
+
+    /**
+     * Invoke the deferred callbacks if the given truth test evaluates to true.
+     *
+     * @param  \Closure  $when
+     * @return void
+     */
+    public function invokeWhen(?Closure $when = null): void
+    {
+        $when ??= fn () => true;
+
+        $this->forgetDuplicates();
+
+        foreach ($this->callbacks as $index => $callback) {
+            if ($when($callback)) {
+                rescue($callback);
+            }
+
+            unset($this->callbacks[$index]);
+        }
+    }
+
+    /**
+     * Remove any deferred callbacks with the given name.
+     *
+     * @param  string  $name
+     * @return void
+     */
+    public function forget(string $name): void
+    {
+        $this->callbacks = collect($this->callbacks)
+            ->reject(fn ($callback) => $callback->name === $name)
+            ->values()
+            ->all();
+    }
+
+    /**
+     * Remove any duplicate callbacks.
+     *
+     * @return $this
+     */
+    protected function forgetDuplicates(): self
+    {
+        $this->callbacks = collect($this->callbacks)
+            ->reverse()
```

Comment:
@taylorotwell not sure, but I think we don't need the double reverse(), right?

Reply:
Thanks